File: ruby/easy/longest_word_in_dictionary.rb
# @param {String[]} words
# @return {String}
def longest_word(words)
    queue = [""]
    longest = ""
    set = Set.new
    visited = Set.new
    words.each do |word|
        set.add(word)
    end
    
    while !queue.empty?
        current = queue.shift
        visited.add(current)
        if current.length > longest.length
            longest = current
        end
        for i in 0..25
            word = current + ('a'.ord + i).chr
            if set.include?(word) && !visited.include?(word)
                queue.push(word)
            end
        end
    end
    return longest
end

File: javascript/easy/longestWordInDictionary.js
/**
 * @param {string[]} words
 * @return {string}
 */
var longestWord = function(words) {
    const queue = new Array(""), visited = new Set(), dict = new Set();
    let longest = "";
    for(let word of words) {
        dict.add(word);
    }
    
    while(queue.length > 0) {
        let current = queue.shift();
        visited.add(current);
        if(current.length > longest.length) {
            longest = current;
        }
        
        for(let i = 0; i < 26; i++) {
            let word = current + String.fromCharCode(i + 97);
            if(dict.has(word) && !visited.has(word)) {
                queue.push(word);
            }
        }
    }
    return longest;
};
